src/Simulation/Simulators/QuantumSimulator/StateDumper.cs

```diff
@@ -155,25 +323,163 @@ public virtual string Format(uint idx, double real, double img)
+            // private bool ShowMeasurementDisplayHistogram { get; set; } = false;
+
+            // /// <summary>
+            // ///     Sets the properties of this display dumper from a given
+            // ///     configuration source.
+            // /// </summary>
+            // public JupyterDisplayDumper Configure(IConfigurationSource configurationSource)
+            // {
+            //     configurationSource
+            //         .ApplyConfiguration<bool>("dump.measurementDisplayHistogram", value => ShowMeasurementDisplayHistogram = value)
+            //         .ApplyConfiguration<bool>("dump.truncateSmallAmplitudes", value => TruncateSmallAmplitudes = value)
+            //         .ApplyConfiguration<double>("dump.truncationThreshold", value => TruncationThreshold = value)
+            //         .ApplyConfiguration<BasisStateLabelingConvention>(
+            //             "dump.basisStateLabelingConvention", value => BasisStateLabelingConvention = value
+            //         );
+            //     return this;
+            // }
+
             /// <summary>
-            /// The callback method. Formats the given state and invokes the <see cref="Channel"/>
+            /// A method to call to output a string representation.
+            /// </summary>
+            public virtual Action<string>? FileWriter { get; }
+
+            /// <summary>
+            ///     Constructs a new display dumper for a given simulator.
+            /// </summary>
+            public DisplayableStateDumper(QuantumSimulator sim, Action<string>? fileWriter = null) : base(sim)
+            {
+                this.FileWriter = fileWriter;
+            }
+
+            /// <summary>
+            ///     Used by the simulator to provide states when dumping.
+            ///     Not intended to be called directly.
             /// </summary>
-            /// <returns>True, so the entire wave function is dumped.</returns>
             public override bool Callback(uint idx, double real, double img)
             {
-                Channel(Format(idx, real, img));
+                if (_data == null) throw new Exception("Expected data buffer to be initialized before callback, but it was null.");
+                _data[idx] = new Complex(real, img);
                 return true;
             }
 
+            /// <summary>
+            ///     Dumps the state of a register of qubits as a Jupyter displayable
+            ///     object.
+            /// </summary>
+            // TODO(rokuzmin): What does it return?
             public override bool Dump(IQArray<Qubit>? qubits = null)
             {
-                var count = qubits == null
-                    ? this.Simulator.QubitManager.AllocatedQubitsCount
-                    : qubits.Length;
-                this._maxCharsStateId = ((1 << (int)count) - 1).ToString().Length;
+                _count = qubits == null
+                            ? this.Simulator.QubitManager?.AllocatedQubitsCount ?? 0
```

Comment:
I assume that with fragment 
> I'm not sure I agree with `?? 0` here, since that would mean `_data` has a length of 1, causing the logic below to necessarily fail

you mean that 
"If at the moment of dumping 
* the `QubitManager` is not yet assigned then it should result in an exception or an assert (a bug in our code is detected);
* none of the qubits is allocated (`0` qubits are allocated) then dumping should still succeed but should display nothing (or a message like "No qubits allocated").

I'm changing to 
```c#
                System.Diagnostics.Debug.Assert(this.Simulator.QubitManager != null,   // Added.
                    "Internal logic error, QubitManager must be assigned");

                _count = qubits == null
                            ? this.Simulator.QubitManager.AllocatedQubitsCount  // Removed `?? 0`.
                            : qubits.Length;
                _data = new Complex[(_count != 0) ? 1 << ((int)_count) : 0]; // `new Complex[0]` if no qubits are allocated.
                var result = base.Dump(qubits);
```
After this, if no qubits are allocated, then a 0-length array is assigned below:
```c#
var state = new DisplayableState
{
    . . .
    Amplitudes = _data,
};
```
and all the subsequent fragments `Amplitudes.Select()` have nothing to select (and return an empty array).

Is my interpretation and fix correct?

Reply:
> the QubitManager is not yet assigned then it should result in an exception or an assert (a bug in our code is detected);

Right, I don't think there's a valid set of conditions under which `QubitManager` has not yet been initialized, but `DumpMachine` has been called by a user. Catching that condition as a bug seems like the most robust thing to do there instead of deferring the problem to later.

> none of the qubits is allocated (0 qubits are allocated) then dumping should still succeed but should display nothing (or a message like "No qubits allocated").

Ah, that's a good point. I guess technically in that case the Hilbert space of the system is ℂ¹ (that is, complex-valued scalars), such that the definition of `_data` below makes sense in that case — namely, as an array of length 1. In particular, `new Complex[1]` makes more sense to me than `new Complex[0]`, since the Hilbert space is one-dimensional rather than zero-dimensional. 

---

Up to a global phase, there's one valid vector in ℂ¹, such that `Amplitudes` should be a length 1 array containing a complex number of the form 𝑒^{𝑖θ} for some θ ∈ ℝ.

Reply:
Thank you for the explanation. Your words make me believe that the line below should stay as it was (if 0 qubits are allocated then the array has a single element)
```c#
_data = new Complex[1 << ((int)_count)];
```

Reply:
I think that's correct, yes. I was under the misunderstanding that if no qubits had been allocated by users, there was still at least one available via the qubit manager. Based on your clarifications above, I think that line is correct as-is. Thanks for your patience!